android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/ConsentManager.java

```diff
@@ -0,0 +1,268 @@
+                .get(subpopulationGuid);
+    }
+
+    // if the participant's session indicates consent to this subpopulation, use that. otherwise,
+    // treat presense of consent in DAO as having consented
+    private boolean isConsentedInSessionOrLocal(UserSessionInfo session, String subpopulationGuid) {
+        ConsentStatus subpopulationStatus = getConsentStatus(subpopulationGuid);
+        if (subpopulationStatus != null && subpopulationStatus.getConsented()) {
+            return true;
+        }
+        return dao.get(subpopulationGuid) != null;
+    }
+
+    /**
+     * @return true if all required consents have been signed
+     */
+    public boolean isConsented() {
+        UserSessionInfo userSessionInfo = authenticationManager.getUserSessionInfo();
+        if (userSessionInfo != null) {
+            if (userSessionInfo.getConsented()) {
+                return true;
+            }
+
+            // Bridge session doesn't specify any consents
+            if (userSessionInfo.getConsentStatuses() == null) {
+                return true;
+            }
+
+            for (Map.Entry<String, ConsentStatus> consentStatus
+                    : userSessionInfo.getConsentStatuses().entrySet()) {
+                if (consentStatus.getValue().getRequired()
+                        && !isConsentedInSessionOrLocal(userSessionInfo, consentStatus.getKey())) {
+                    return false;
+                }
+            }
+            return true;
+        }
+        // without a user session, we can't determine whether they're consented. this shouldn't
+        // happen unless they're logged out of the application
+        return false;
+    }
+
+    /**
+     * @return true if all *required* consents have been signed and the versions signed are the most
+     * up-to-date versions of those consents
+     */
+    public boolean isConsentedMostRecent() {
+        UserSessionInfo userSessionInfo = authenticationManager.getUserSessionInfo();
+        return userSessionInfo == null ? false : userSessionInfo.getSignedMostRecentConsent();
+    }
+
+    /**
+     * @param subpopulationGuid guid for the subpopulation of the consent
+     * @param name              participant's full name
+     * @param birthdate         participant's date of birth
+     * @param base64Image       participant's signature, encoded
+     * @param imageMimeType     mime type of participant's signature
+     * @param sharingScope      participant's sharing scope for the study
+     * @return completable
+     */
+    @NonNull
+    public Completable giveConsent(@NonNull String subpopulationGuid, @NonNull String name,
+                                   @NonNull LocalDate birthdate,
+                                   @NonNull String base64Image, @NonNull String imageMimeType,
+                                   @Nullable SharingScope sharingScope) {
+        checkNotNull(subpopulationGuid);
+        checkNotNull(name);
```

Comment:
Interesting. It appears Bridge  requires name, bday, and sharingScope but not imageData or mimeType. 

It also appears that Bridge checks to make sure you're 18, so I'm thinking I'd want to enforce that in android-sdk before letting it go into local storage... 

This also means that what we thought was a valid consent locally is rejected by Bridge, which I had not considered. This adds some complexity since we can't even actually call Bridge to give consent prior to creating the account... so right now it's possible to do consent workflow (saved offline), sign up, consent gets uploaded in background, and is rejected by Bridge.. user now is unconsented